src/background_music/Integrator.h

```diff
@@ -0,0 +1,222 @@
+#ifndef AUDACIOUS_PLUGINS_INTEGRATOR_H
+#define AUDACIOUS_PLUGINS_INTEGRATOR_H
+/*
+ * RC-circuit style integrator.
+ * Copyright 2023 Michel Fleur
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions, and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions, and the following disclaimer in the documentation
+ *    provided with the distribution.
+ *
+ * This software is provided "as is" and without any warranty, express or
+ * implied. In no event shall the authors be liable for any damages arising from
+ * the use of this software.
+ */
+#include "utils.h"
+#include <cmath>
+#include <limits>
+#include <stdexcept>
+#include <type_traits>
+
+template<typename S>
+class Integrator
+{
+    static_assert(std::is_floating_point_v<S>);
+
+    S history_multiplier_;
+    S input_multiplier_;
+
+public:
+    /**
+     * Below this fraction of a sample the input multiplier cannot be
+     * distinguished from 1.0 and the history multiplier would become de-normal.
+     */
+    static S minimum_samples()
+    {
+        static S min = -1.0 / (std::log(std::numeric_limits<S>::epsilon()));
+        return min;
+    }
+
+    /**
+     * Above this number of samples, the history multiplier cannot be
+     * distinguished from 1.0, which leads to unstable results.
+     */
+    static S maximum_samples()
+    {
+        static S max = -1.0 / std::log(1.0 - std::numeric_limits<S>::epsilon());
+        return max;
+    }
+
+    /**
+     * Returns the history multiplier for integration over the provided number
+     * of samples, that can be a fraction. If the fraction is smaller than
+     * {@code minimum_samples}, zero is returned. If the number is bigger than
+     * {@code maximum_samples} the history multiplier for {@code
+     * maximum_samples} is returned.
+     * @param samples The number of samples
+     * @return the history multiplier that lies between [0 .. 1>.
+     */
+    static inline S calculate_history_multiplier_from_samples(S samples)
+    {
+        S absolute = fabs(samples);
+        return absolute < minimum_samples()   ? 0
+               : absolute > maximum_samples() ? exp(-1 / maximum_samples())
+                                            : exp(-1.0 / absolute);
+    }
+
+    /**
+     * Calculates the history multiplier from a given input multiplier and vice
+     * versa.
+     * @param multiplier The multiplier
+     * @return The other multiplier
+     */
+    [[maybe_unused]] static inline S calculate_other_multiplier(S multiplier)
+    {
+        return 1 - multiplier;
```

Comment:
We don't need to define a new function for an operation so simple as "one minus x". Doing so makes the code less readable, not more.